pkg/schema/type.go

```diff
@@ -66,101 +72,107 @@ func (m ScalarType) GetValueType() yamlmeta.Type {
 	panic("Not implemented because it is unreachable")
 }
 
-func (t *DocumentType) CheckType(_ yamlmeta.TypeWithValues, _ string) (chk yamlmeta.TypeCheck) {
+func (t *DocumentType) CheckType(_ yamlmeta.TypeWithValues) (chk yamlmeta.TypeCheck) {
 	return
 }
-func (m *MapType) CheckType(node yamlmeta.TypeWithValues, prependErrorMessage string) (chk yamlmeta.TypeCheck) {
-	violationErrorMessage := prependErrorMessage + " was type %T when %T was expected"
-
+func (m *MapType) CheckType(node yamlmeta.TypeWithValues) (chk yamlmeta.TypeCheck) {
 	nodeMap, ok := node.(*yamlmeta.Map)
 	if !ok {
-		scalar, ok := node.(*yamlmeta.Scalar)
-		if ok {
-			chk.Violations = append(chk.Violations, fmt.Sprintf(violationErrorMessage, scalar.Value, &yamlmeta.Map{}))
-		} else {
-			chk.Violations = append(chk.Violations, fmt.Sprintf(violationErrorMessage, node, &yamlmeta.Map{}))
-		}
+		chk.Violations = append(chk.Violations,
+			NewTypeError(node.ValueTypeAsString(), typeToString(m), node.GetPosition(), m.Position))
 		return
 	}
 
 	for _, item := range nodeMap.Items {
 		if !m.AllowsKey(item.Key) {
-			chk.Violations = append(chk.Violations, fmt.Sprintf("Map item '%s' at %s is not defined in schema", item.Key, item.Position.AsCompactString()))
+			chk.Violations = append(chk.Violations,
+				NewUnexpectedKeyError(item.Key.(string), item.Position, m.Position))
 		}
 	}
 	return
 }
-func (t MapItemType) CheckType(node yamlmeta.TypeWithValues, prependErrorMessage string) (chk yamlmeta.TypeCheck) {
-	violationErrorMessage := prependErrorMessage + " was type %T when %T was expected"
-
+func (t *MapItemType) CheckType(node yamlmeta.TypeWithValues) (chk yamlmeta.TypeCheck) {
 	mapItem, ok := node.(*yamlmeta.MapItem)
 	if !ok {
-		chk.Violations = append(chk.Violations, fmt.Sprintf(violationErrorMessage, node.GetValues()[0], node))
-		return
+		// A Map must've yielded a non-MapItem which is not valid YAML
+		log.Panicf("MapItem type check was called on a non-MapItem: %#v", node)
```

Comment:
should be using plain `panic(...)` for internal consistency checks.

Reply:
Updated to use regular panic instead of log.Panic'ing to remove redundant output